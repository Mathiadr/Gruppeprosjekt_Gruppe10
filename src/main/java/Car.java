import java.util.ArrayList;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    private String availableDate;   //TODO: IMPLEMENT
    private boolean isAvailable = true; //Whether or not the car is occupied (Not to be confused with available date)
                                        //TODO: IMPLEMENT
    private String fuelType;        //TODO: IMPLEMENT
    private String transmission;    //TODO: IMPLEMENT

    private ArrayList<String> features;


    private static ArrayList<Car> carArrayList = new ArrayList<>();

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
    }


    public String getAvailableDate() {
        return availableDate;
    }

    public void setAvailableDate(String availableDate) {
        this.availableDate = availableDate;
    }

    public static ArrayList<Car> getCarArrayList() {
        return carArrayList;
    }

    public static void setCarArrayList(ArrayList<Car> carArrayList) {
        Car.carArrayList = carArrayList;
    }

    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }
}
