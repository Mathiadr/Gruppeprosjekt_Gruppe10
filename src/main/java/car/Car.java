package car;

import java.util.ArrayList;
import java.util.Collection;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    //private String availableDate;   //TODO: IMPLEMENT
    private boolean isAvailable = true; //Whether the car is occupied (Not to be confused with available date)
                                        //TODO: IMPLEMENT
    //private String fuelType;        //TODO: IMPLEMENT
    //private String transmission;    //TODO: IMPLEMENT

    //private Collection<String> features; //TODO: IMPLEMENT

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }


    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }
}
