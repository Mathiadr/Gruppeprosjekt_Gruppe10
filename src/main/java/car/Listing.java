package car;

import java.util.Date;
import java.time.format.DateTimeFormatter;

public class Listing {
        // Contains information of every bid, each of which belongs to one car.
        // TODO: EVERY CAR BELONGS TO ONE LISTING
        // TODO: ur mum
        private String publisherUsername;
        private Date startDate;
        private Date endDate;
        private boolean isAvailable;
        private boolean published;
        private String description;


        public Listing(){}

        public Listing(String publisherUsername, Date startDate, Date endDate, boolean published, String description){
            this.publisherUsername = publisherUsername;
            this.startDate = startDate;
            this.endDate = endDate;
            this.isAvailable = true;
            this.published = published;
            this.description = description;


        }


        public Date getStartDate() {
            return startDate;
        }

        public void setStartDate(Date startDate) {
            this.startDate = startDate;
        }

        public Date getEndDate() {
            return endDate;
        }

        public void setEndDate(Date endDate) {
            this.endDate = endDate;
        }

        public boolean isAvailable() {
            return isAvailable;
        }

        public void setAvailable(boolean available) {
            isAvailable = available;
        }

        public boolean isPublished() {
            return published;
        }

        public void setPublished(boolean published) {
            this.published = published;
        }
}

