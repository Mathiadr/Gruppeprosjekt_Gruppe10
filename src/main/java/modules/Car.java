package modules;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Date;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    private String brand;                   //TODO: IMPLEMENT
    //private String fuelType;              //TODO: IMPLEMENT
    //private String transmission;          //TODO: IMPLEMENT
    //private Collection<String> features;  //TODO: IMPLEMENT
    private Listing listing;

    public static final String[] fuelTypesList = {"Electric", "Gasoline", "Diesel"};
    public static final String[] transmissionList = {"Manual", "Automatic"};
    public static final String[] featureList = {"GPS", "Air Conditioning"};



    public Car(){

    }

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
        LocalDate startDateplaceholder = LocalDate.of(2022, 1, 1);
        LocalDate endDatePlaceholder = LocalDate.of(2022, 12, 31);
        this.listing = new Listing(startDateplaceholder, endDatePlaceholder, true, "");
    }
    public Car(String registrationNumber, String owner, String model, Listing listing) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
        this.listing = listing;
    }


    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public Listing getListing() {
        return listing;
    }

    public void setListing(Listing listing) {
        this.listing = listing;
    }



    @Override
    public String toString(){
        return "\n\nregistrationNumber: " + registrationNumber + ",\n" +
                "Owner: " + owner + ",\n" +
                "Model: " + model + ",\n" +
                "Listing: " + listing;
    }
}
