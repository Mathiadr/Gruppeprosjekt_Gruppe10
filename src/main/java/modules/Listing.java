package modules;


import java.util.Date;

public class Listing {
    // Contains information of every bid, each of which belongs to one car.
    // TODO: EVERY CAR BELONGS TO ONE LISTING
    // TODO: ur mum
    private String publisherUsername;
    private Date startDate;
    private Date endDate;
    private Car car;
    private boolean isAvailable;
    private boolean published;
    private String description;

    public Listing(){}

    public Listing(String publisherUsername, Date startDate, Date endDate, Car car, boolean published, String description){
        this.publisherUsername = publisherUsername;
        this.startDate = startDate;
        this.endDate = endDate;
        this.car    = car;
        this.isAvailable = true;
        this.published = published;
        this.description = description;
    }


    public Date getStartDate() {
        return startDate;
    }

    public void setStartDate(Date startDate) {
        this.startDate = startDate;
    }

    public Date getEndDate() {
        return endDate;
    }

    public void setEndDate(Date endDate) {
        this.endDate = endDate;
    }

    public Car getCar() {
        return car;
    }

    public void setCar(Car car) {
        this.car = car;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }

    public boolean isPublished() {
        return published;
    }

    public void setPublished(boolean published) {
        this.published = published;
    }



}
