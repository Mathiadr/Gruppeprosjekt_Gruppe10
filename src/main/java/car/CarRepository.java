package car;

import org.json.JSONObject;

import java.io.*;
import java.util.ArrayList;

public class CarRepository {
    private String repositoryName;

    private static ArrayList<Car> availableCarArrayList; // Holds cars not currently occupied
    private static ArrayList<Car> allCarArrayList; // Holds ALL cars. Unsure if necessary

    CarRepository(String repositoryName) throws FileNotFoundException {
        this.repositoryName = repositoryName;

        FileReader fileReader = new FileReader(this.repositoryName);
    }

    // Adds a new car to the repository file
    public void AddNewCar(Car car){
        //TODO: implement
        JSONObject jsonObject = new JSONObject(car.getRegistrationNumber());
        jsonObject.put("registrationNumber", car.getRegistrationNumber());
        jsonObject.put("owner", car.getOwner());
        jsonObject.put("model", car.getModel());
        jsonObject.put("availableDate", car.getAvailableDate());
        jsonObject.put("isAvailable", car.isAvailable());
        jsonObject.put("fuelType", car.getFuelType());
        jsonObject.put("transmission", car.getTransmission());
        jsonObject.put("features", car.getFeatures()); //Maybe a nested JSONObject/JSONArray belongs here?
        try{
            FileWriter fileWriter = new FileWriter(this.repositoryName);
            fileWriter.append(jsonObject.toString());
            System.out.println("Car " + car.getRegistrationNumber() + " has been written to file.");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // It rents da car, dumbass
    public void RentCar(Car car){
        //TODO: implement
    }

    // Returns all cars that are not currently occupied
    public ArrayList<Car> GetAllAvailableCars(){
        //TODO: implement
        return null;
    }

    // Unsure what to do here...
    public void SaveAllAvailableCars(){
        //TODO: implement
    }


}
