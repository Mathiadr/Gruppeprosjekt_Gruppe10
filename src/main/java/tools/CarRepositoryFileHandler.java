package tools;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import interfaces.FileHandler;
import modules.Car;


import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

public class CarRepositoryFileHandler implements FileHandler<Car> {

    @Override
    public ArrayList<Car> readFromFile(String filename) {
        ObjectMapper objectMapper = new ObjectMapper();
        File file = new File(filename);
        try {
            return objectMapper.readValue(file, new TypeReference<ArrayList<Car>>() {});
        } catch (IOException e) {
            System.err.println(e);
        }
        return null;
    }

    @Override
    public void writeArrayListToFile(ArrayList<Car> carArrayList, String filename) {
        ObjectMapper objectMapper = new ObjectMapper();
        File file = new File(filename);

        try {
            objectMapper.writerWithDefaultPrettyPrinter().writeValue(file, carArrayList);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}