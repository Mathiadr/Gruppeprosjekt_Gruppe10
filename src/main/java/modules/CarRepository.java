package modules;

import tools.CarRepositoryFileHandler;

import java.io.File;
import java.util.ArrayList;

public class CarRepository {
    private String repositoryName; // This holds the name of the repository being accessed.
    private ArrayList<Car> carArrayList = new ArrayList<>();
    private final CarRepositoryFileHandler carRepositoryFileHandler = new CarRepositoryFileHandler();

    public CarRepository(String repositoryName) {
        this.repositoryName = repositoryName;

        carArrayList = readFromJSON();
    }



    // Adds a new car to the repository file
    public void AddNewCar(Car car){
        //TODO: implement
        boolean exists = false;
        for (Car i : carArrayList){
            //Checks if the car being added does not already exist within the list
            if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                exists = true;
                break;
            }
        }
        if (!exists){
            //System.out.println("Adding car " + car.getRegistrationNumber() + " with car owner " + car.getOwner() + " to carArrayList");
            this.carArrayList.add(car);
        }
        else{
            //System.out.println("Car " + car.getRegistrationNumber() + " Already exists");
        }
    }

    // Removes Car object from carArrayList
    public void RemoveExistingCar(Car car){
        if (carArrayList.contains(car)){
            //System.out.println("Removed car " + car.getRegistrationNumber() + " with car owner " + car.getOwner() + " from carArrayList");
            carArrayList.remove(car);
        }
        else{
            // TODO: Implement error message to GUI
            //System.out.println("Could not find car " + car.getRegistrationNumber() + " in CarArrayList");
        }
    }

    // It rents da car, dumbass
    public void RentCar(Car car){
        // TODO: Test if works
        try{
            for (Car i : GetAllAvailableCars()){
                if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                    //TODO: ADD EXTRA MEASURE TO ENSURE IT DOES NOT RENT AN UNAVAILABLE CAR
                    i.getListing().setAvailable(false);
                }
            }
        }
        catch (RuntimeException couldNotRentCar){
            //System.out.println("\n\nCould not rent car " + car.getRegistrationNumber() + "\n\n");
        }

    }

    // Returns all cars that are not currently occupied
    public ArrayList<Car> GetAllAvailableCars(){
        ArrayList<Car> allAvailableCars = new ArrayList<>();
        for (Car i : carArrayList){
            if (i.getListing().isAvailable()){
                allAvailableCars.add(i);
            }
        }
        return allAvailableCars;
    }

    public void updateListing(Car car, Listing listing){
        for (Car i : carArrayList){
            if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                i.setListing(listing);
                //System.out.println("Updated listing in Car " + i.getRegistrationNumber());
                break;
            }
        }
    }

    public void SaveCarsToJSON(){
        carRepositoryFileHandler.writeArrayListToFile(getCarArrayList(), repositoryName);
    }

    // Reads all objects from JSON file and maps it to the Car object
    public ArrayList<Car> readFromJSON(){
        return carRepositoryFileHandler.readFromFile(repositoryName);
    }

    public String getRepositoryName() {
        return repositoryName;
    }

    public void setRepositoryName(String repositoryName) {
        this.repositoryName = repositoryName;
    }

    public ArrayList<Car> getCarArrayList() {
        return carArrayList;
    }

    public void setCarArrayList(ArrayList<Car> carArrayList) {
        this.carArrayList = carArrayList;
    }
}
