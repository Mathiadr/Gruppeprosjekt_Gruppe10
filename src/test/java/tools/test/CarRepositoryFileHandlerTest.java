package tools.test;

import modules.Car;
import modules.CarRepository;
import modules.Listing;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import tools.CarRepositoryFileHandler;

import java.io.IOException;
import java.time.LocalDate;
import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;

public class CarRepositoryFileHandlerTest {
    CarRepository carRepository;
    CarRepositoryFileHandler carRepositoryFileHandler;

    @BeforeAll
    static void initializeCarRepository(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        CarRepositoryFileHandler carRepositoryFileHandler = new CarRepositoryFileHandler();
        carRepository.addNewCar(
                new Car("MXH10045", "Mathias Dale Ratdal", "E2",
                        new Listing(
                                LocalDate.of(2022, 1, 1),
                                LocalDate.of(2022, 3, 7),
                                true, "PLACEHOLDER DESCRIPTION")));
        carRepository.addNewCar(
                new Car("EBX23365", "Sokrat Starynkiewicz", "A4",
                        new Listing(
                                LocalDate.of(2022, 2, 25),
                                LocalDate.of(2022, 6, 7),
                                true, "PLACEHOLDER DESCRIPTION")));
        carRepository.addNewCar(
                new Car("E0B10045", "Joe Mama", "C2",
                        new Listing(
                                LocalDate.of(2022, 2, 1),
                                LocalDate.of(2022, 2, 10),
                                true, "PLACEHOLDER DESCRIPTION")));
        carRepository.addNewCar(
                new Car("ADHD0O11", "Sander Kristiansen", "C6",
                        new Listing(
                                LocalDate.of(2022, 7, 15),
                                LocalDate.of(2022, 8, 15),
                                true, "PLACEHOLDER DESCRIPTION")));
        carRepository.addNewCar(
                new Car("D1CK9864", "Djikstra Van Houven", "BA1",
                        new Listing(
                                LocalDate.of(2022, 10, 14),
                                LocalDate.of(2022, 11, 4),
                                true, "PLACEHOLDER DESCRIPTION")));
        carRepository.addNewCar(
                new Car("P0043456", "John Doe", "A2",
                        new Listing(
                                LocalDate.of(2022, 12, 1),
                                LocalDate.of(2022, 12, 31),
                                true, "PLACEHOLDER DESCRIPTION")));
        carRepository.addNewCar(
                new Car("BE240051", "Ola Nordmann", "A1",
                        new Listing(
                                LocalDate.of(2022, 2, 7),
                                LocalDate.of(2022, 12, 23),
                                true, "PLACEHOLDER DESCRIPTION")));

        carRepository.saveCarsToJSON();
        carRepository.readFromJSON();
    }

    @BeforeEach
    void resetCarRepository(){
        carRepository = new CarRepository("testRepository.json");
        carRepository.readFromJSON();
    }

    @Test
    /*
     *  Tester krav:
     *      500. Databasen skal oppbevare informasjon om registrerte biler
     *      501. Databasen skal oppbevare bilmerke
     *      502. Databasen skal oppbevare navn p책 bilens eier
     *      600. Databasen skal oppbevare annonsen opprettet av utleier
     *      604. Databasen skal oppbevare tidsperiode for n책r bilen er tilgjengelig for utleie
     */
    public void repository_is_persistently_stored_in_file(){
        Car newCar = new Car("DFG441563", "Dummy one", "Volvo");
        carRepository.addNewCar(newCar);
        carRepository.saveCarsToJSON();
        carRepository.readFromJSON();
        Approvals.verify(carRepository.getCarArrayList());
    }

    @Disabled("Could not manage to get test to work. Kept to document attempt.")
    @Test
    public void writing_to_file_throws_exception_if_invalid(){
        String filename = "WrongFilename.txt";
        IOException ioException = assertThrows(IOException.class,
                () -> carRepositoryFileHandler.writeArrayListToFile(carRepository.getCarArrayList(), filename));
        assertEquals("", ioException.getMessage());
    }

    @Disabled("Could not manage to get test to work. Kept to document attempt.")
    @Test
    public void writing_to_file_throws_exception_if_arrayList_is_empty(){
        ArrayList<Car> carArrayList = new ArrayList<>(); // Empty
        NullPointerException nullPointerException =
                assertThrows(NullPointerException.class,
                () -> carRepositoryFileHandler.writeArrayListToFile(carArrayList, carRepository.getRepositoryName()));
        assertEquals("", nullPointerException.getMessage());
    }

    @Test
    /*
     *  Tester krav:
     *      500. Databasen skal oppbevare informasjon om registrerte biler
     *      501. Databasen skal oppbevare bilmerke
     *      502. Databasen skal oppbevare navn p책 bilens eier
     *      600. Databasen skal oppbevare annonsen opprettet av utleier
     *      604. Databasen skal oppbevare tidsperiode for n책r bilen er tilgjengelig for utleie
     */
    public void repository_reads_from_file(){
        carRepository.readFromJSON();
        Approvals.verify(carRepository.getCarArrayList());
    }
}
