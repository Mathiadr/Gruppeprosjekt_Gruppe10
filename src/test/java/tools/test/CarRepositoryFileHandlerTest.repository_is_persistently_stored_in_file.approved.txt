[

registrationNumber: MXH10045,
Owner: Mathias Dale Ratdal,
Model: E2,
Listing:    Start date: 2022-01-01,
   End date: 2022-03-07,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: EBX23365,
Owner: Sokrat Starynkiewicz,
Model: A4,
Listing:    Start date: 2022-02-25,
   End date: 2022-06-07,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: E0B10045,
Owner: Joe Mama,
Model: C2,
Listing:    Start date: 2022-02-01,
   End date: 2022-02-10,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: ADHD0O11,
Owner: Sander Kristiansen,
Model: C6,
Listing:    Start date: 2022-07-15,
   End date: 2022-08-15,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: D1CK9864,
Owner: Djikstra Van Houven,
Model: BA1,
Listing:    Start date: 2022-10-14,
   End date: 2022-11-04,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: P0043456,
Owner: John Doe,
Model: A2,
Listing:    Start date: 2022-12-01,
   End date: 2022-12-31,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: BE240051,
Owner: Ola Nordmann,
Model: A1,
Listing:    Start date: 2022-02-07,
   End date: 2022-12-23,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: DFG441563,
Owner: Dummy one,
Model: Volvo,
Listing:    Start date: 1900-01-01,
   End date: 1900-01-01,
   isAvailable: true,
   Description : ]